### PR TITLE
Add Tag and Revision support to waypoints, gateway API.

### PR DESCRIPTION
**Pull request by therealmitchconnors**
Needs default tag handling and tests.

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 1]
```diff
@@ -237,7 +249,9 @@ func NewDeploymentController(client kube.Client, clusterID cluster.ID,
 	// On injection template change, requeue all gateways
 	injectionHandler(func() {
 		for _, gw := range dc.gateways.List(metav1.NamespaceAll, klabels.Everything()) {
-			dc.queue.AddObject(gw)
+			if string(gw.Spec.GatewayClassName) == gw.GetName() {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:252]: why are we comparing the GWC to the GW name?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 2]
```diff
@@ -271,6 +285,15 @@ func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
 		if !knownControllers.Contains(string(gc.Spec.ControllerName)) {
 			return nil
 		}
+		// find the tag or revision indicated by the object
+		// TODO: handle default tag
+		selectedTag, ok := gw.Labels[label.IoIstioRev.Name]
+		if !ok {
+			selectedTag = d.namespaces.Get(gw.Namespace, "").Labels[label.IoIstioRev.Name]
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:292]: `d.namespaces.Get(gw.Namespace, "")` can be nil

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 3]
```diff
@@ -271,6 +285,15 @@ func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
...
+		}
+		if !d.myTags.Contains(selectedTag) {
+			return nil
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:295]: we will need to make the leader election smarter as well I think, right now only 1 pod will even run this controller at a time

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 4]
```diff
@@ -412,6 +435,16 @@ func (d *DeploymentController) apply(controller string, yml string) error {
 	return nil
 }
 
+func (d *DeploymentController) HandleTagChange(newTags []string) {
+	d.myTags = sets.Set[string]{}
+	d.myTags.InsertAll(newTags...)
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:440]: nit: `sets.New(newTags...)`

File: pkg/revisions/tag_watcher.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+func NewTagWatcher(client kube.Client, revision string) TagWatcher {
+	p := &tagWatcher{
+		revision:        revision,
+		mu:              sync.RWMutex{},
+		tagsToRevisions: map[string]string{},
+		revisionsToTags: map[string][]string{},
+	}
+	p.queue = controllers.NewQueue("tag", controllers.WithReconciler(p.updateTags))
+	p.webhookInformer = client.KubeInformer().Admissionregistration().V1().MutatingWebhookConfigurations().Informer()
```
Comment (howardjohn) [inline pkg/revisions/tag_watcher.go:61]: nit: can use the new kclient

Comment (howardjohn) [reply pkg/revisions/tag_watcher.go:61]: Actually its not really a nit, HasSynced is currently incorrect without it

File: pkg/revisions/tag_watcher.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+func (p *tagWatcher) HasSynced() bool {
+	return p.queue.HasSynced()
+}
+
+// notifyHandlers notifies all registered handlers on tag change.
+// assumes externally locked.
+func (p *tagWatcher) notifyHandlers() {
+	for _, handler := range p.handlers {
+		handler(append(p.revisionsToTags[p.revision], p.revision))
```
Comment (howardjohn) [inline pkg/revisions/tag_watcher.go:92]: This will mutate `p.revisionsToTags[p.revision]` sometimes

File: pkg/revisions/tag_watcher.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	"istio.io/api/label"
+	"istio.io/istio/istioctl/pkg/tag"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/pkg/log"
+)
+
+// TagWatcher keeps track of the current tags and can notify watchers
+// when the tags change.
+type TagWatcher interface {
```
Comment (howardjohn) [inline pkg/revisions/tag_watcher.go:33]: We should probably make the defaulthandler just a syntax on top of this perhaps. Not a blocker though

**Review by howardjohn — COMMENTED** [review]
Overall concept LGTM

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 1 above)
Comment (therealmitchconnors) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:252]: This was copy pasta.  Fixing now.

File: pkg/revisions/tag_watcher.go (hunk 5 above)
Comment (therealmitchconnors) [reply pkg/revisions/tag_watcher.go:61]: In that case, we should update default_watcher, which is the source of this code...

File: pkg/revisions/tag_watcher.go (hunk 6 above)
Comment (therealmitchconnors) [reply pkg/revisions/tag_watcher.go:92]: I can't see it.  Not saying you're wrong, but I will need this one spelled out for me...

File: pilot/pkg/bootstrap/configcontroller.go [hunk 8]
```diff
@@ -174,14 +175,17 @@ func (s *Server) initK8SConfigStore(args *PilotArgs) error {
...
+							tagWatcher := revisions.NewTagWatcher(s.kubeClient, args.Revision)
 							controller := gateway.NewDeploymentController(s.kubeClient, s.clusterID, s.webhookInfo.getWebhookConfig, s.webhookInfo.addHandler)
+							tagWatcher.AddHandler(controller.HandleTagChange)
 							// Start informers again. This fixes the case where informers for namespace do not start,
 							// as we create them only after acquiring the leader lock
 							// Note: stop here should be the overall pilot stop, NOT the leader election stop. We are
 							// basically lazy loading the informer, if we stop it when we lose the lock we will never
 							// recreate it again.
 							s.kubeClient.RunAndWait(stop)
 							controller.Run(leaderStop)
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/configcontroller.go:187]: Run() blocks so L188 will never run I think

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:241]: Given https://github.com/istio/istio/pull/44174 this should actually be remove leader election I think.

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 9]
```diff
@@ -339,6 +364,15 @@ func (d *DeploymentController) apply(controller string, yml string) error {
 	return nil
 }
 
+func (d *DeploymentController) HandleTagChange(newTags []string) {
+	d.myTags = sets.New(newTags...)
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:368]: This is racy w/o mutex.

Personally I would just pass the entire TagWatcher into DeploymentController, have it expose a `watcher.GetTags() sets.String`, and use that everywhere; should simplify things a little bit.

You still need HandleTagChange to enqueue things, though

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 10]
```diff
@@ -339,6 +364,15 @@ func (d *DeploymentController) apply(controller string, yml string) error {
...
+	for _, gw := range d.gateways.List(metav1.NamespaceAll, klabels.Everything()) {
+		if string(gw.Spec.GatewayClassName) == gw.GetName() {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:370]: GatewayClassName will never == gw.GetName (unless its a coincidence)

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 9 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:368]: Then `Run` also doesn't need to take it as a param

File: pkg/revisions/tag_watcher.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+	p := &tagWatcher{
+		revision:        revision,
+		mu:              sync.RWMutex{},
+		tagsToRevisions: map[string]string{},
+		revisionsToTags: map[string][]string{},
+	}
+	p.queue = controllers.NewQueue("tag", controllers.WithReconciler(p.updateTags))
+	p.webhookInformer = kclient.NewFiltered[*admissionregistrationv1.MutatingWebhookConfiguration](client, kubetypes.Filter{
+		ObjectFilter:    isTagWebhook,
+		ObjectTransform: kube.StripUnusedFields,
```
Comment (howardjohn) [inline pkg/revisions/tag_watcher.go:64]: nit: this is default, not needed

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 10 above)
Comment (therealmitchconnors) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:370]: I fixed that once.  It must have crept back into the PR when I switched dev environments...

File: pkg/revisions/tag_watcher.go (hunk 11 above)
Comment (therealmitchconnors) [reply pkg/revisions/tag_watcher.go:64]: Can you clarify what is default?  The UI is showing four lines of code, and I am fairly confident some of them are not default...

Comment (howardjohn) [reply pkg/revisions/tag_watcher.go:64]: kube.StripUnusedFields

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 12]
```diff
@@ -215,6 +232,20 @@ func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
 		if !knownControllers.Contains(string(gc.Spec.ControllerName)) {
 			return nil
 		}
+		// find the tag or revision indicated by the object
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:235]: this should be outside the L230 `if`; there are other paths that allow getting to configureIstioGateway

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 13]
```diff
@@ -215,6 +232,20 @@ func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
...
+		selectedTag, ok := gw.Labels[label.IoIstioRev.Name]
+		if !ok {
+			ns := d.namespaces.Get(gw.Namespace, "")
+			if ns == nil {
+				return nil
+			}
+			selectedTag = ns.Labels[label.IoIstioRev.Name]
+		}
+		myTags := d.tagWatcher.GetMyTags()
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:244]: do we need to consider the label on the Gateway itself?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 14]
```diff
@@ -339,6 +370,13 @@ func (d *DeploymentController) apply(controller string, yml string) error {
 	return nil
 }
 
+func (d *DeploymentController) HandleTagChange(newTags sets.Set[string]) {
+	log.Warnf("Handling set of tags %v", newTags)
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:374]: nit: not a warning

File: pkg/revisions/tag_watcher.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+		log.Errorf("failed to sync tag watcher")
+		return
+	}
+
+	p.queue.Run(stopCh)
+}
+
+// AddHandler registers a new handler for updates to tag changes.
+func (p *tagWatcher) AddHandler(handler TagHandler) {
+	p.mu.Lock()
```
Comment (howardjohn) [inline pkg/revisions/tag_watcher.go:85]: We have inconsistent locking. Here we lock to add handlers, but later we call notifyHandlers without a lock (despite the function comment saying it assumes a lock). We should do one or the other (probably "with lock" is best)

File: pkg/revisions/tag_watcher.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+	wh := p.webhookInformer.Get(key.Name, "")
+	p.webhookInformer.Get(key.Name, "")
+	if wh != nil {
+		revision = wh.GetLabels()[label.IoIstioRev.Name]
+		tagName = wh.GetLabels()[tag.IstioTagLabel]
+	}
+	p.mu.Lock()
+	defer func() {
+		p.mu.Unlock()
+		p.notifyHandlers()
```
Comment (howardjohn) [inline pkg/revisions/tag_watcher.go:132]: this is the part I meant about notifyHandlers not being under a lock

File: pkg/revisions/tag_watcher.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+	myTags := p.GetMyTags()
+	for _, handler := range p.handlers {
+		handler(myTags)
+	}
+}
+
+func (p *tagWatcher) updateTags(key types.NamespacedName) error {
+	var revision, tagName string
+	wh := p.webhookInformer.Get(key.Name, "")
+	p.webhookInformer.Get(key.Name, "")
```
Comment (howardjohn) [inline pkg/revisions/tag_watcher.go:124]:
```suggestion
```

File: pkg/revisions/tag_watcher.go (hunk 16 above)
Comment (therealmitchconnors) [reply pkg/revisions/tag_watcher.go:132]: notifyHandlers calls GetMyTags, which acquires a readLock, so I will need to move locking logic into notifyHandlers.

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 12 above)
Comment (therealmitchconnors) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:237]: I'm not sure I follow.  If the gateway is of the correct revision, but not the correct class, we shouldn't reconcile it, right?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 13 above)
Comment (therealmitchconnors) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:246]: I think we're doing that at L238

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:248]: Will be a followup PR

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 12 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:237]: L253 allows for it to be the correct class as well

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 18]
```diff
@@ -190,10 +204,13 @@ func NewDeploymentController(client kube.Client, clusterID cluster.ID,
...
 
+	dc.tagWatcher.AddHandler(dc.HandleTagChange)
+
 	return dc
 }
 
 func (d *DeploymentController) Run(stop <-chan struct{}) {
+	kube.WaitForCacheSync(stop, d.tagWatcher.HasSynced)
 	d.queue.Run(stop)
 	controllers.ShutdownAll(d.deployments, d.services, d.serviceAccounts, d.gateways, d.gatewayClasses)
```
Comment (zhlsunshine) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:215]: is there a missing for **d.namespaces** which need to be shutdown as well?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 19]
```diff
@@ -223,6 +240,21 @@ func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
...
+	selectedTag, ok := gw.Labels[label.IoIstioRev.Name]
+	if !ok {
+		ns := d.namespaces.Get(gw.Namespace, "")
+		if ns == nil {
+			return nil
+		}
+		selectedTag = ns.Labels[label.IoIstioRev.Name]
+	}
+	myTags := d.tagWatcher.GetMyTags()
+	if !myTags.Contains(selectedTag) && !(selectedTag == "" && myTags.Contains("default")) {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:253]: I am not sure I follow this. Can you expand a bit on the concern?

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:253]: NVM, this is right

**Pull request merged**